src/v3.rs

```diff
@@ -339,11 +340,45 @@ impl Personalization {
 
     /// Add a dynamic template data field.
     pub fn add_dynamic_template_data(mut self, dynamic_template_data: SGMap) -> Personalization {
-        self.dynamic_template_data
-            .get_or_insert_with(|| SGMap::with_capacity(dynamic_template_data.len()))
-            .extend(dynamic_template_data);
+        // We can safely unwrap & unreachable here since SGMap will always serialize
+        // to a JSON object.
+        let mut new_vals = match to_value(dynamic_template_data).unwrap() {
+            Object(map) => map,
+            _ => unreachable!(),
+        };
+        if let Some(mut old_vals) = self.dynamic_template_data {
+            old_vals.append(&mut new_vals);
+            self.dynamic_template_data = Some(old_vals);
+        } else {
+            self.dynamic_template_data = Some(new_vals);
+        }
         self
     }
+
+    /// Add a dynamic template data fields from a json object.
+    pub fn add_dynamic_template_data_json<T: Serialize + ?Sized>(
+        mut self,
+        json_object: &T,
+    ) -> SendgridResult<Personalization> {
+        let mut new_vals = match to_value(json_object)? {
+            Object(map) => map,
+            _ => {
+                return Err(SendgridError::RequestNotSuccessful(
```

Comment:
this was the only error in `SendgridError` that I could create, not sure if there is a better way of going about this.

adding to the enum felt more breaking but if that is cleaner happy to do so

Reply:
Maybe we could use the [`JSONDecode`](https://docs.rs/sendgrid/0.16.0/sendgrid/error/enum.SendgridError.html#variant.JSONDecode) variant here?

Reply:
As far as I can tell you cannot externally construct [serde_json::error::Error](https://docs.rs/serde_json/1.0.59/serde_json/error/struct.Error.html).

One option is to construct an `io::Error` which `serde_json::error::Error` has a from defined for, however that would introduce a new dependency into the library.

Reply:
Ah, good point. I forgot that was a wrapper around a crate error. I am fine with introducing a new variant here. Since it's a new function, I don't think it would be a breaking change.

Reply:
added a new variant!